core/src/main/java/bisq/core/offer/BsqSwapWalletWatcher.java

```diff
@@ -24,86 +24,131 @@
+                onOpenOffersRemoved(c.getRemoved());
+            }
+        };
     }
 
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // API
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
     public void onAllServicesInitialized() {
         btcWalletService.addChangeEventListener(this);
         bsqWalletService.addChangeEventListener(this);
+        openOfferManager.getObservableList().addListener(listChangeListener);
+        onOpenOffersAdded(openOfferManager.getObservableList());
     }
 
     public void shutDown() {
         btcWalletService.removeChangeEventListener(this);
         bsqWalletService.removeChangeEventListener(this);
+        openOfferManager.getObservableList().removeListener(listChangeListener);
     }
 
 
     ///////////////////////////////////////////////////////////////////////////////////////////
-    // API
+    // WalletChangeEventListener
     ///////////////////////////////////////////////////////////////////////////////////////////
 
-    public void addListener(Listener listener) {
-        if (!listeners.contains(listener) && listener.getOffer().isBsqSwapOffer()) {
-            listeners.add(listener);
-        }
+    public void onWalletChanged(Wallet wallet) {
+        openBsqSwapOffers.forEach(this::updateFundingState);
     }
 
-    // Remove listeners by Offer
-    public void removeListener(Offer offer) {
-        var toRemove = listeners.stream()
-                .filter(listener -> listener.getOffer().equals(offer))
-                .collect(Collectors.toList());
-        toRemove.forEach(listeners::remove);
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // Private
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    private void onOpenOffersAdded(List<? extends OpenOffer> list) {
+        list.stream().filter(openOffer -> openOffer.getOffer().isBsqSwapOffer())
+                .forEach(openOffer -> {
+                    openBsqSwapOffers.add(openOffer);
+                    updateFundingState(openOffer);
+                });
+    }
+
+    private void onOpenOffersRemoved(List<? extends OpenOffer> list) {
+        list.stream().filter(openOffer -> openOffer.getOffer().isBsqSwapOffer())
+                .forEach(openBsqSwapOffers::remove);
     }
 
-    public boolean isFunded(Offer offer) {
-        if (!feeService.isFeeAvailable())
-            return false;
+    private void updateFundingState(OpenOffer openOffer) {
+        if (!feeService.isFeeAvailable()) {
+            // At startup the fee service might not be available still so we call again after a delay
+            UserThread.runAfter(() -> {
+                if (openBsqSwapOffers.contains(openOffer)) {
+                    updateFundingState(openOffer);
+                }
+            }, 10);
+            return;
+        }
```

Comment:
Looks like this would create an infinite loop if `isFeeAvailable()` keeps being false.

Reply:
Yes, it repeats as long the fees are not available after every 10 sec. I think thats ok, but could add some counter as well or increase delay with each iteration.

Reply:
It's doing it for every offer which is a bit unnecessary, would be better to have this loop not for each offer but just once for the instance

Reply:
yes i added a counter to reduce dependency on fee service. we have default rate if service is failing